package/MDAnalysis/analysis/waterdynamics.py

```diff
@@ -1208,87 +1208,84 @@ class SurvivalProbability(object):
         self.t0 = t0
         self.tf = tf
         self.dtmax = dtmax
-        self.nproc = nproc
-        self.timeseries = None
+        self.timeseries = []
 
-    def _getOneDeltaPoint(self, selection, totalFrames, t0, tau):
-        """
-        Gives one point to calculate the mean and gets one point of the plot
-        C_vect vs t.
 
-        - Ex: t0=1 and tau=1 so calculate the t0-tau=1-2 interval
-        - Ex: t0=5 and tau=3 so calculate the t0-tau=5-8 interval
-        """
-        Ntau = self._NumPart_tau(selection, totalFrames, t0, tau)
-        Nt = float(self._NumPart(selection, t0))
-        return Ntau/Nt if Nt > 0 else 0
+    def run(self, **kwargs):
+        """Analyze trajectory and produce timeseries"""
 
-    def _getMeanOnePoint(self, universe, selection1, selection_str, wint,
-                         totalFrames):
+        # select all frames to an array
+        selected = self._selection_serial(self.universe, self.selection)
+
+        if len(selected) < self.dtmax:
+            print ("ERROR: Cannot select fewer frames than dtmax")
+            return
+
+        for window_size in list(range(1, self.dtmax + 1)):
+            output = self._getMeanOnePoint(selected, window_size)
+            self.timeseries.append(output)
+
+
+    def _selection_serial(self, universe, selection_str):
+        selected = []
+        pm = ProgressMeter(self.tf-self.t0, interval=10,
+                           verbose=True, offset=-self.t0)
+        for ts in universe.trajectory[self.t0:self.tf]:
+            selected.append(universe.select_atoms(selection_str))
+            pm.echo(ts.frame)
+        return selected
+
+
+    def _getMeanOnePoint(self, selected, window_size):
         """
         This function gets one point of the plot P(t) vs t. It uses the
         _getOneDeltaPoint() function to calculate the average.
-
         """
-        n = 0.0
+        n = 0
         sumDeltaP = 0.0
-        for frame in range(totalFrames - wint):
-            a = self._getOneDeltaPoint(selection1, totalFrames, frame, wint)
-            sumDeltaP += a
+        for frame_no in range(len(selected) - window_size):
+            delta = self._getOneDeltaPoint(selected, frame_no, window_size)
+            sumDeltaP += delta
             n += 1
 
-        # if no water molecules remain in selection, there is nothing to get
-        # the mean, so n = 0.
-        return sumDeltaP/n if n > 0 else 0
+        return sumDeltaP/n
 
-    def _NumPart_tau(self, selection, totalFrames, t0, tau):
+
+    def _getOneDeltaPoint(self, selected, t, tau):
         """
-        Compares the molecules in t0 selection and t0+tau selection and
-        select only the particles that remain from t0 to t0+tau. It returns
-        the number of remaining particles.
+        Gives one point to calculate the mean and
+        gets one point of the plot C_vect vs t.
+        - Ex: t0=1 and tau=1 so calculate
+        how many selected beads survive from the frame 1 to 2
```

Comment:
I am changing it back to "water molecules" to make it consistent. The reason why I changed it is because the function can be used for anything in the solution or any other type of solution.